### DISCUSSION TITLE
Gokrazy API

### DISCUSSION THREAD

**Discussion by BrunoTeixeira1996** [gokrazy/gokrazy#229] [Q&A] [answered]
Hi guys,
I was wondering if there is a way to list (in a easy way) all gokrazy api endpoints?
I wanted to extract some info to use in another source and I need to use files for that (i.e config.json).
So my question is, is there a way to curl gokrazy device and extract some info? like gokrazy token

**Comment by damdo** [comment, marked as answer]
Hey Bruno!
As far as I know (and @stapelberg can correct me here) there is not yet a full documentation on the gokrazy on-device API endpoints.

But you should be able to find most of them by using grep or a similar tool, and then read the relevant code.
Here is what I found with a trivial search:
```
$ rg "http.HandleFunc"
status.go
208:    http.HandleFunc("/status", func(w http.ResponseWriter, r *http.Request) {
277:    http.HandleFunc("/log", func(w http.ResponseWriter, r *http.Request) {
329:    http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {

supervise.go
698:    http.HandleFunc("/stop", stopstartHandler)
699:    http.HandleFunc("/restart", stopstartHandler)

update.go
286:    http.HandleFunc("/update/features", func(w http.ResponseWriter, r *http.Request) {
310:    http.HandleFunc("/update/mbr", nonConcurrentUpdateHandler(rootdev.BlockDevice()))
311:    http.HandleFunc("/update/root", nonConcurrentUpdateHandler(rootdev.Partition(rootdev.InactiveRootPartition())))
312:    http.HandleFunc("/update/switch", nonConcurrentSwitchHandler(rootdev.InactiveRootPartition()))
313:    http.HandleFunc("/update/testboot", nonConcurrentTestbootHandler(rootdev.InactiveRootPartition()))
316:            http.HandleFunc(fmt.Sprintf("/update/device-specific/%s", extraUpdateHandler.name), nonConcurrentLimitedUpdateHandler(extraUpdateHandler.device, extraUpdateHandler.offset, extraUpdateHandler.maxLength))
322:    http.HandleFunc("/update/boot", updateHandler)
323:    http.HandleFunc("/update/bootonly", func(w http.ResponseWriter, r *http.Request) {
331:    http.HandleFunc("/reboot", func(w http.ResponseWriter, r *http.Request) {
382:    http.HandleFunc("/poweroff", func(w http.ResponseWriter, r *http.Request) {
433:    http.HandleFunc("/uploadtemp/", uploadTemp)
434:    http.HandleFunc("/divert", divert)
```
That said I think long term the intention is to centralise the API specification to a central place for all these types, which we are already using for GUS.

To your other question, at the moment I don't think you can read the entire `config.json` from an endpoint.
For now you can get the SBOM Hash, which is a SHA265 hash of the `config.json` file used to generate the gokrazy instance, by performing a GET to `/` and parsing the JSON output.

But I guess in the future we could implement something to return the `config.json` too, by either adding it to what is returned by `/` or by adding a special endpoint, or even by adding a way to read any file from the device through the API (like for example Talos Linux [does](https://www.talos.dev/v1.5/reference/cli/#talosctl-read)).

**Reply by BrunoTeixeira1996** [reply]
Thanks for the detailed response! and for the heads up!
Btw whats the `rg` you are using? something like `grep -rn` ?

**Reply by stapelberg** [reply]
rg is ripgrep.

The JSON output that gokrazy provides (when the `Accept: application/json` header is present) on the `/` and `/status` handlers is quite detailed: 

```
curl --header 'Accept: application/json' http://gokrazy:secret@router7/
```

Maybe this is already good enough for what you’re planning to do? You didn’t share many details about your use-case, so it’s hard to say.
